apps/settings/js/modules/base/dependency_graph.js

```diff
@@ -0,0 +1,134 @@
+define(function(require) {
+  'use strict';
+
+  var Module = require('modules/base/module');
+
+  /**
+   * DependencyGraph is a basic data structure that helps decide the evaluation
+   * order of a certain node.
+   *
+   * @class DependencyGraph
+   * @requires module:modules/base/module
+   * @params {DependencyGraph} dpGraph
+   *                           The newly created graph will be initialized using
+   *                           dpGraph.
+   * @returns {DependencyGraph}
+   */
+  var DependencyGraph = Module.create(function DependencyGraph(dpGraph) {
+    this._nodes = dpGraph ? JSON.parse(JSON.stringify(dpGraph._nodes)) : {};
+  });
+
+  /**
+   * Print the content of the nodes for debugging.
+   *
+   * @access private
+   * @memberOf DependencyGraph.prototype
+   */
+  DependencyGraph.prototype._printNodes = function dg_printNodes() {
+    Object.keys(this._nodes).forEach((name) => {
+      var node = this._nodes[name];
+      console.log('name: ' + name);
+      console.log('children: ' +
+      node.children.reduce((result, name) => {
+        return result + ' ' + name;
+      }, ''));
+      console.log('dependent nodes: ' +
+      node.dependentNodes.reduce((result, name) => {
+        return result + ' ' + name;
+      }, ''));
+      console.log('======================');
+    });
+  };
+
+  /**
+   * Traverse the tree and add the dependent nodes to all nodes in the path.
+   *
+   * @access private
+   * @memberOf DependencyGraph.prototype
+   * @param {Object} node
+   *                 The node that we have new dependent nodes to add to.
+   * @param {Array} dependentNodes
+   *                New dependent nodes to add.
+   */
+  DependencyGraph.prototype._addDependentNodes =
+    function dg_addDependentNodes(node, dependentNodes, traversedNodes) {
+      traversedNodes = traversedNodes || [];
+      if (traversedNodes.indexOf(node.name) >= 0) {
+        traversedNodes.push(node.name);
+        var error = traversedNodes.join(' -> ');
+        this.Error('circular dependency detected! ' + error);
```

Comment:
Because in your previous revision, `this.Error` refers to `console.log()`, I think we should return after this line to make sure we won't do following works. But now, you would throw error in this case, I think your approach do fix the problem I said.